source/server/overload_manager_impl.cc

```diff
@@ -266,32 +274,54 @@ ThreadLocalOverloadState& OverloadManagerImpl::getThreadLocalOverloadState() {
   return tls_->getTyped<ThreadLocalOverloadStateImpl>();
 }
 
-void OverloadManagerImpl::updateResourcePressure(const std::string& resource, double pressure) {
-  auto action_range = resource_to_actions_.equal_range(resource);
-  std::for_each(action_range.first, action_range.second,
-                [&](ResourceToActionMap::value_type& entry) {
-                  const std::string& action = entry.second;
-                  auto action_it = actions_.find(action);
-                  ASSERT(action_it != actions_.end());
-                  const OverloadActionState old_state = action_it->second.getState();
-                  if (action_it->second.updateResourcePressure(resource, pressure)) {
-                    const auto state = action_it->second.getState();
-
-                    if (old_state.isSaturated() != state.isSaturated()) {
-                      ENVOY_LOG(debug, "Overload action {} became {}", action,
-                                (state.isSaturated() ? "saturated" : "scaling"));
-                    }
-                    tls_->runOnAllThreads([this, action, state] {
-                      tls_->getTyped<ThreadLocalOverloadStateImpl>().setState(action, state);
-                    });
-                    auto callback_range = action_to_callbacks_.equal_range(action);
-                    std::for_each(callback_range.first, callback_range.second,
-                                  [&](ActionToCallbackMap::value_type& cb_entry) {
-                                    auto& cb = cb_entry.second;
-                                    cb.dispatcher_.post([&, state]() { cb.callback_(state); });
-                                  });
-                  }
-                });
+void OverloadManagerImpl::updateResourcePressure(const std::string& resource, double pressure,
+                                                 int flush_epoch) {
+  auto [start, end] = resource_to_actions_.equal_range(resource);
+
+  std::for_each(start, end, [&](ResourceToActionMap::value_type& entry) {
+    const std::string& action = entry.second;
+    auto action_it = actions_.find(action);
+    ASSERT(action_it != actions_.end());
+    const OverloadActionState old_state = action_it->second.getState();
+    if (action_it->second.updateResourcePressure(resource, pressure)) {
+      const auto state = action_it->second.getState();
+
+      if (old_state.isSaturated() != state.isSaturated()) {
+        ENVOY_LOG(debug, "Overload action {} became {}", action,
+                  (state.isSaturated() ? "saturated" : "scaling"));
+      }
+      state_updates_to_flush_.insert_or_assign(action, state);
```

Comment:
The calls to insert_or_assign in this function mean we could lose some updates but I think that is ok. Should we add unit tests for this behavior?

Reply:
Yep, done.

Reply:
Can you add some comments here for the next reader?

Reply:
Sorry I meant the part about potentially losing updates.